### ISSUE TITLE
CP Subsystem resilience to restarts and auto-promotion to CP membership

### ISSUE THREAD

**Issue by roytmana** [hazelcast/hazelcast#20385]
When CP system membership is formed on startup a predefined number of nodes are elected to CP members. as those members fail or restarted ( a common case in containerized environment) they get new identity and do not join back to CP. Any non CP nodes would not join to CP automatically either and thus CP membership declines overtime until CP system fails irrevocable and need to be re-formed. This is a one way downhill road that will guarantee failure unless something intervenes.

While there are APIs to promote a node to CP (or remove failed node from CP) maintaining CP membership it is a non trivial matter and best left to cluster itself. Manual management is possible but should be a last resort

I was experimenting with some (naive) solutions while learning hazelcast (I am a newbie) I tried to decrease CP auto removal timeout to 60 seconds and I make all my nodes to be promoted to CP on startup. But short timeout makes it very possible old node that was auto-removed will try to join back (any safeguards for it like treating it as new?) and CP promotion on startup may create too many CP nodes (not in my case but in general) and may fail on startup and will need to be retried later complicating my application logic and it is not applicable to standalone hazelcast distributions only to embedded. I also tried a scenario where I allocated more CP members that I (my CP group) needs to have a safety net. It worked but it means that all CP members must be available to form the initial CP membership 

One possible approach would be to make CP membership a little more elastic by introducing concept of CP eligible members vs CP active members. I have a 5 nodes 3 CP members + 2 CP eligible nodes. If an active member leaves cluster (shutdown) or times out (separate from auto removal timeout) a CP Eligible member is brought into CP membership. if timed out old node rejoins  before auto-removal it can stay (we will have larger number of members) or it can be demoted to CP Eligible. Technically every node could be considered CP eligible so I am not sure there is a need to have this distinction as long as there is a mechanism to promote available nodes into CP to maintain its membership counts 
A simple initial/temporary/compromise solution could to allow forming CP with a specified **minimum** number of nodes but greedily bring up to **maximum** nodes  into CP if available including on startup any new node (if we are below maximum). This will give some safety net with more CP nodes than my groups need acting like standby with little change to existing logic keeping CP system further away from complete failure and giving more chance to recover with homegrown solutions (like I can check on node startup if cluster is below maximum and promote starting node to CP without having to promote all of them)

I also wonder if there is any value in supporting CP-only nodes - a stripped down build of hazelcast which is only serving CP needs as a low processor/memory deployment. In my current application I (will) have 2-3 worker nodes with embedded hazelcast doing my processing and I probably end up having 3 more light nodes (out of the box hazelcast or embedded hazelcast-only nodes if I need custom logic like CP promotion on startup) just to form a stable CP system as the basis so my worker nodes can start and stop without fear of breaking entire CP system. These light nodes will be sharing resources with other things so the lighter they are the better. But maybe hazelcast is very efficient and does not need it. I use elasticsearch extensively and it has a concept of roles for each node (multiple per node) which control tasks that node performs in a cluster ...

**Comment by roytmana** [comment]
@jerrinot as discussed on slack

**Comment by mmedenjak** [comment]
Thank you for the suggestions @roytmana . 

@vbekiaris @petrpleshachkov @sancar @arodionov can you take a look at this issue and [the other one](https://github.com/hazelcast/hazelcast/issues/20319) and provide some thoughts?

**Comment by Holmistr** [comment]
@arodionov Just a quick thought - would this be something that is solved when you have EE with CP Persistence enabled?

**Comment by arodionov** [comment]
@Holmistr yes, you are right. Without state persistence, the CP subsystems could only tolerate/recover after the network partitioning because all its states are in memory.
Regarding the proposed solution: first of all, it is should be verified and aligned to the Raft protocol and should be proven that it couldn't bring the cluster to an inconsistent state.
